### PR TITLE
fix(ast): Fix AST bug

### PR DESCRIPTION
**Pull request by ragaeeb**
# GPT Codex 5.3

Implemented both fixes in [src/App.tsx](src/App.tsx).

1. Preview spinner stuck forever
- Root cause: `FilePreview` depends on `onLoadContent`, and `loadFileContent` was recreated each render. That caused effect cleanup cancellation before `isLoading` could be reset.
- Fix: made `loadFileContent` stable with `useCallback` at [src/App.tsx:397](src/App.tsx:397).

2. Bug report download not saving
- Fix: changed bug report export to the same pattern as pack “Export All” flow:
  - pick a folder via `open({ directory: true })`
  - `authorize_export_directory`
  - create filename and `join(folder, filename)`
  - write via `write_file_content`
- Updated at [src/App.tsx:593](src/App.tsx:593) and imports at [src/App.tsx:1](src/App.tsx:1).

Validation run:
- `bun test` passed.
- `bun run vite:build` passed.

Note: your working tree already had unrelated modifications (`package.json`, `tsconfig.json`, `tsconfig.node.json`, `src-tauri/Cargo.lock`) before this; I only changed `src/App.tsx`.

Applied a second round of fixes for both issues.

Changes made:

1. Preview spinner deadlock fix
- Reworked preview load effect to remove the `inFlight` cancellation pattern that can get stuck in React dev/effect double-run.
- `FilePreview` now:
  - sets loading false on cache hit,
  - starts one load path per effect run,
  - only clears spinner from the currently alive run.
- File: [FilePreview.tsx](src/components/FilePreview.tsx)

2. Bug report export flow switched back to save-file dialog
- Reverted from folder picker to `save(...)` dialog.
- Uses same write sequence as the working single-pack save:
  - `save` -> `dirname` -> `authorize_export_directory` -> `write_file_content`.
- File: [App.tsx](src/App.tsx)

3. Added debug logging for diagnosis
- Preview logs now include start/cache-hit/done/error.
- Bug report export logs now include start/cancel/success/fail.
- Preview file read logs now include start/success/fail from app side.
- Files: [App.tsx](src/App.tsx), [FilePreview.tsx](src/components/FilePreview.tsx)

Validation:
- `bun test` passed.
- `bun run vite:build` passed.

Please retry and export logs again.  
If spinner still hangs, send the lines containing `preview-read` and `[preview]` from the generated bug report; those should pinpoint whether the Tauri read is hanging or failing fast.

Implemented both requests.

- Removed the per-pack individual `Save` action from the output panel. Only `Export All` remains in [OutputPreview.tsx](src/components/OutputPreview.tsx:63) and [OutputPreview.tsx](src/components/OutputPreview.tsx:276).
- Fixed the AST dead-code bug that was dropping legitimate TSX code:
  - `.tsx` now uses the TSX grammar (not TS) in [ast.rs](src-tauri/src/commands/ast.rs:5).
  - Reachability now includes JSX identifiers in reference collection in [ast.rs](src-tauri/src/commands/ast.rs:111).
  - BFS is now seeded from entry-file references (important when entry file has no top-level declarations) in [ast.rs](src-tauri/src/commands/ast.rs:252).
  - Frontend stripper now preserves `export default Identifier;` symbols in [ast-reachability.ts](src/lib/ast-reachability.ts:16).

TDD/regression coverage added:

- New frontend tests for default-export identifier preservation:
  - [ast-reachability.test.ts](src/lib/ast-reachability.test.ts:97)
  - [ast-reachability.test.ts](src/lib/ast-reachability.test.ts:108)
- New backend integration-style AST test reproducing your `main/App/LimitIndicator` graph:
  - [ast.rs](src-tauri/src/commands/ast.rs:442)

Validation run:

- `bun test` passed (`130 pass, 0 fail`)
- `cargo test --manifest-path src-tauri/Cargo.toml` passed (`68 pass, 0 fail`)
- `bun run vite:build` passed

Next step:
1. Re-test your exact `main.tsx` + `App.tsx` + `LimitIndicator.tsx` case with AST enabled and entry point set to `main.tsx`.

Implemented all 4 requested changes.

1. Out-of-date pack UX
- Added pack-input fingerprinting and stale detection in [App.tsx](src/App.tsx#L67).
- When selection/options/model change after a pack, the button now:
  - switches to amber,
  - changes label to `Re-Pack N Files`,
  - shows an explicit “Output is out of date” warning.
- Relevant logic/UI: [App.tsx](src/App.tsx#L353), [App.tsx](src/App.tsx#L1022).

2. Removed `Copy + Prompt` and `Copy Pack`
- Removed both copy actions and prompt builder UI from output preview.
- `Export`/`Export All` remains.
- Updated in [OutputPreview.tsx](src/components/OutputPreview.tsx#L52).

3. Num-packs default behavior (1..5 => max, >5 => 5)
- Added defaulting behavior:
  - 1 selected file => slider auto-defaults to 1
  - 5 selected files => auto-defaults to 5
  - 6+ selected files => auto-defaults to 5
- Added manual override behavior: once user changes the slider, auto-defaulting stops (except hard cap).
- Logic in [App.tsx](src/App.tsx#L713), [App.tsx](src/App.tsx#L997).

4. New dev doc for macOS build/sign/notarize
- Added [macos-build-signing.md](docs/macos-build-signing.md).
- Includes:
  - prerequisites,
  - cert identity setup,
  - notarization env options (API key or Apple ID/app-specific password),
  - repo-accurate build commands (`bun run tauri build --target ...`),
  - verification and troubleshooting commands.

Also updated README feature bullets for current behavior in [README.md](README.md#L20).

Validation run:
- `bun test` passed.
- `bun run vite:build` passed.

Online references used for “up to date” check:
- [Tauri v2 prerequisites](https://v2.tauri.app/start/prerequisites/)
- [Tauri v2 macOS signing/notarization](https://v2.tauri.app/distribute/sign/macos/)
- [Tauri CLI build targets (universal-apple-darwin)](https://docs.rs/crate/tauri-cli/latest/source/README.md)
- [Apple app-specific passwords](https://support.apple.com/en-ca/102654)

## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive macOS build & signing guide for Tauri v2.
  * README updated to remove individual-pack export option (now "Export all packs in one folder").

* **New Features**
  * Pack out-of-date indicator with Re‑Pack workflow when selection/options change.
  * App-level logging with selectable log levels and debug event capture for previews and packs.

* **Improvements**
  * Simplified export UX to a single Export All control.
  * Updated file icon colors and menu/app naming to improve clarity.

* **Chores**
  * CI workflows and package metadata updated; stricter TypeScript compiler settings.

### CODE REVIEW

**Comment by ragaeeb** [comment]
# V2 Review & Next Steps (Gemini 3.1 Pro)

Here is a review of your outgoing changes, brainstorming of missed edge cases, and suggestions for testing.

## ✅ What We Fixed (Good Stuff)

1. **AST JSX Parsing Correctness**:
   - Switching `tsx` to use `LANGUAGE_TSX` (instead of `LANGUAGE_TYPESCRIPT`) is a massive fix. Previously, the TS parser would choke on JSX tags like `<div>` and treat them as boolean operator errors.
   - Adding `jsx_identifier` to `node.kind()` checks successfully allows component references (like `<LimitIndicator />`) to be registered in the reference graph.

2. **Reachability Bootstrap via Entry Point**:
   - The original Rust reachability algorithm only started from "symbols declared in the entry point". For codebases where the entry point is an unnamed script (like `createRoot(<App/>)`), the queue never started. 
   - By seeding the AST Reachability queue with `file_level_refs` of the entry point, the dependency graph will correctly execute down through imported components.

3. **Output UX Polish**:
   - Stripped the clunky prompt builder and replaced it with a dynamic "Export / Export All" workflow directly tied to the length of the `packResult`. 

## 🚨 Bugs & Missed Edge-Cases

### 1. AST Dead-Code Ignores Unused *Exported* Symbols
**File:** `src/lib/ast-reachability.ts` (lines 33-35)
Currently, `stripUnreachableSymbols` completely skips stripping a symbol if `isLikelyExportedSymbol` returns `true`. 
**Impact:** This defeats a huge purpose of cross-file reachability analysis! If the Rust backend correctly marks `export function deadUtility()` as unreachable across all files, the frontend just returns it untouched to avoid string-replacement headaches (like leaving a dangling `export ` keyword). Massive token waste occurs because unused exports are kept.
**How to Fix:** Provide a more robust AST-to-Source mapper (e.g., passing node start/end byte offsets from Rust to TS) to safely strip the *entire* exported node rather than relying on regex.

### 2. Export Default Regex is Brittle to Comments
**File:** `src/lib/ast-reachability.ts`
You added: `new RegExp('^\\s*export\\s+default\\s+${escaped}\\s*;?\\s*$', "m")`
This regex terminates on `$`. If a user writes:
```tsx
export default App; // The main app shell
```
The regex fails. Because it fails `isLikelyExportedSymbol`, the string stripper proceeds and will obliterate `const App = ...`, but leave the orphaned `export default App; // The main app shell` floating at the bottom of the file. 
**How to Fix:** Relax the line-ending validation to allow comments/trailing characters:
`new RegExp('^\\s*export\\s+default\\s+${escaped}\\b', "m")`

### 3. Orphaned/Dangling Re-exports
**File:** `src-tauri/src/commands/ast.rs`
If you `export { Button } from "./Button"`, `Button` is highly unlikely to be declared in the intermediate file, but acts as a pass-through. The Rust AST parser doesn't correlate exported external module aliases back to the origin module unless it traces `export ... from`. If the entry point hits the intermediary, it might not find the final `Button` implementation. 

## 🧪 Suggested Test Coverage Additions

To catch regressions and improve robustness, I recommend adding these tests:

### 1. Trailing Comment Export Cases
In `ast-reachability.test.ts`, add:
```typescript
it("should identify default exports with trailing comments", () => {
  const input = `const X = 1;\nexport default X; // some comment`;
  const output = stripUnreachableSymbols(input, ["X"], "js");
  expect(output).toContain("export default X");
});
```

### 2. Integration: The Cross-File Unused Export Check
You should have an E2E or CLI workflow test verifying that unused exports *are* actually dropped across files:
```typescript
it("drops unused exported utilities", () => {
    // If File A exports UtilityX and UtilityY, and App only uses UtilityX...
    // Verify UtilityY is missing from the final pack content.
});
```
*(Currently this test would fail, which proves the global reachability limitation mentioned in #1)*

### 3. Empty Edge-Cases for `OutputPreview.tsx`
Add a component test for `OutputPreview` when zero packs are generated, ensuring no index out-of-bounds on tabs or crash on export (though Tauri prevents this, it guarantees robust component UI state). Use React Testing Library to verify the "Export" button doesn't blow up on an empty `packResult.packs` array.

File: src/App.tsx [hunk 1]
```diff
@@ -394,24 +445,48 @@ export default function App() {
...
+            return next.length > 5000 ? next.slice(next.length - 5000) : next;
+          });
+        }
+      } catch {
+        setFileContents((prev) => {
+          if (prev.has(path)) return prev;
+          const next = new Map(prev);
+          next.set(path, "");
+          return next;
+        });
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply src/App.tsx]: https://github.com/ragaeeb/bablusheed/pull/14/commits/ccc8984aa1fb72adf5eb4f48f6906279feb086d3
